src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+                {
+                    end = forStatementSyntax.Declaration.FullSpan.End;
+                }
+                else
+                {
+                    // Should not be reachable because we returned earlier for this case
+                    end = 0;
+                    Debug.Fail("Complete statement not supported on For statement iterator");
+                }
+            }
+            else
+            {
+                end = currentNode.Span.End;
+            }
+
+            // If a node's semicolon is missing, the trailing trivia (including new line) is associated with the last token.
+            // To avoid placing the semicolon on the next line or after comments, we need to adjust the position.
+            return root.FindToken(end).GetPreviousToken().Span.End;
+        }
+
+        private static bool CaretIsInForStatementDeclaration(int caretPosition, ForStatementSyntax forStatementSyntax)
+        {
+            return caretPosition > forStatementSyntax.Declaration.Span.Start && caretPosition < forStatementSyntax.Declaration.Span.End;
+        }
+
+        private static bool CaretIsInForStatementCondition(int caretPosition, ForStatementSyntax forStatementSyntax)
+        {
+            return caretPosition > forStatementSyntax.Condition.SpanStart && caretPosition < forStatementSyntax.Condition.Span.End;
+        }
+
+        /// <summary>
+        /// Examines the enclosing statement-like syntax for an expression which is eligible for statement completion.
+        /// </summary>
+        /// <remarks>
+        /// <para>This method tries to identify <paramref name="currentNode"/> as a node located within an argument
+        /// list, where the immediately-containing statement resembles an "expression statement". This method returns
+        /// <see langword="true"/> if the node matches a recognizable pattern of this form.</para>
+        /// </remarks>
+        private static bool LooksLikeNodeInArgumentList(SyntaxNode currentNode,
+            int caret, bool isCaretAtEndOfLine, ISyntaxFactsService syntaxFacts)
+        {
+            // work our way up the tree, looking for a node of interest within the current statement
+            bool nodeFound = false;
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                // Look for the following types of nodes:
+                // ArgumentList: covers method invocations like object.Method(arg)
+                // ArrayRankSpecifier: covers new Type[dim]
+                // ElementAccessExpression: covers indexer invocations like array[index]
+                // ParenthesizedExpression: covers (3*(x+y))
+                if (currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier, SyntaxKind.ElementAccessExpression, SyntaxKind.ParenthesizedExpression))
+                {
+                    nodeFound = true;
```

Comment:
why not just break here.  and then check if we're in a Statement/Field-decl below?    Needing this specialized local, and needing null checks below just makes this super confusing to understand the flow around.

Reply:
It wasn't readily working so I filed a follow-up issue to address it.  https://github.com/dotnet/roslyn/issues/32260